src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/FallbackProjectManager.cs

```diff
@@ -0,0 +1,163 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.IO;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+/// <summary>
+/// This class is responsible for maintaining project information for projects that don't
+/// use the Razor or Web SDK, or otherwise don't get picked up by our CPS bits, but have
+/// .razor or .cshtml files regardless.
+/// </summary>
+[Shared]
+[Export(typeof(FallbackProjectManager))]
+internal sealed class FallbackProjectManager
+{
+    private readonly ProjectConfigurationFilePathStore _projectConfigurationFilePathStore;
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+    private readonly ProjectSnapshotManagerAccessor _projectSnapshotManagerAccessor;
+
+    private ImmutableHashSet<ProjectId> _fallbackProjectIds = ImmutableHashSet<ProjectId>.Empty;
+
+    [ImportingConstructor]
+    public FallbackProjectManager(
+        ProjectConfigurationFilePathStore projectConfigurationFilePathStore,
+        LanguageServerFeatureOptions languageServerFeatureOptions,
+        ProjectSnapshotManagerAccessor projectSnapshotManagerAccessor)
+    {
+        _projectConfigurationFilePathStore = projectConfigurationFilePathStore;
+        _languageServerFeatureOptions = languageServerFeatureOptions;
+        _projectSnapshotManagerAccessor = projectSnapshotManagerAccessor;
```

Comment:
One thing I've noticed is that Razor tooling code often takes a `ProjectSnapshotManagerAccesor` and holds onto it to access the instance later. Would it be better to just hold onto the result of the `Instance` property? It feels like that might make code using a `ProjectSnapshotManager` cleaner.

Reply:
You're just going to make me air all of our dirty laundry in public, aren't you? 😛

I think this is a case where consistency wins over perfection. There are two completely independent project systems running in Razor. One of them has the lifetime of Visual Studio, one of them has a lifetime controlled by the LSP client (essentially the lifetime of an open solution). In the former, you'll always get the same instance. In the latter, making sure to always get the right instance is very important. Also some classes function in both systems.

For this class specifically it would not be a problem since its on the right side of the fence (it will end up calling here: https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioWindowsProjectSnapshotManagerAccessor.cs#L27) but I would worry about someone copying that pattern.

Of course in The Long Term™ this all goes away anyway.

Reply:
> You're just going to make me air all of our dirty laundry in public, aren't you? 😛

Yup! :smile: `ProjectSnapshotManagerAccessor` has inherent bad smell of an architecture that has been retrofitted, rather than designed.

Reply:
For sure. Not sure if its because of needing the dispatcher, so direct access wasn't supposed to be easy, or if the DI used in the langauge server doesn't support Lazy like MEF does, but its definitely odd. Looking forward to deleting it :D

Reply:
> I think this is a case where consistency wins over perfection. There are two completely independent project systems running in Razor. One of them has the lifetime of the Roslyn VisualStudioWorkspace, one of them has a lifetime controlled by the LSP client (essentially the lifetime of an open solution). In the former, you'll always get the same instance. In the latter, making sure to always get the right instance is very important. Also some classes function in both systems.

I took a closer look today, and both of the implementations calculate the `Instance` property once and return the same value with each access. What am I missing?

Reply:
Knowing you, I would venture to guess that you are missing nothing, and either i'm wrong, or the code has changed sufficiently that past assumptions don't hold.

Reply:
I'm guessing it's the latter. It might not be needed anymore due to past churn.